kedro-datasets/kedro_datasets/partitions/partitioned_dataset.py

```diff
@@ -300,7 +311,32 @@ def _load(self) -> dict[str, Callable[[], Any]]:
 
         return partitions
 
+    async def _async_load(self) -> dict[str, Callable[[], Any]]:
+        partitions = {}
+
+        async def load_partition(partition: str) -> None:
+            kwargs = deepcopy(self._dataset_config)
+            kwargs[self._filepath_arg] = self._join_protocol(partition)
+            dataset = self._dataset_type(**kwargs)  # type: ignore
+            partition_id = self._path_to_partition(partition)
+            partitions[partition_id] = dataset.load
+
+        await asyncio.gather(
+            *[load_partition(partition) for partition in self._list_partitions()]
+        )
```

Comment:
If I understand correctly, there's no actual I/O being performed here, right? Only the `partitions` dictionary is being populated.

I don't see the need of using async helpers and `asyncio.gather` here.

If anything, as a user I'd expect to have the `async` loaders available in my node function so that I can `await` them (provided that my node is asynchronous), use `asyncio.gather` myself, or use an `asyncio.TaskGroup`.

```yaml
my_partitioned_dataset:
  type: partitions.PartitionedDataset
  path: s3://my-bucket-name/path/to/folder
  ...
  use_async: True
```

```python
def concat_partitions(partitioned_input: dict[str, Awaitable]) -> pd.DataFrame:
    tasks = []
    async with asyncio.TaskGroup() as tg:
        for partition_key, partition_load_func in sorted(partitioned_input.items()):
            tasks.append(tg.create_task(partition_load_func()))

    result = pd.DataFrame()
    result = pd.concat([result] + [tasks.result() for task in tasks], ignore_index=True, sort=True)
```

(not that I find this a particularly friendly DX, but it's more or less a continuation of our current approach https://docs.kedro.org/en/stable/data/partitioned_and_incremental_datasets.html#partitioned-dataset-load)

What am I missing?

Reply:
That makes sense. I am open to both the options. Let me know if you want to revert the load method to the original definition. Happy to also update the documentation once we are aligned with the changes made

Reply:
Having fine control of async in node is ok IMO. Is there a way of avoiding coupling node logic to it?

IIRC the original question was using the async option of Runner, and we found that the partitioned dataset only do async on the whole dataset level.

Reply:
Is there any reason why we prefer making it at the dataset level rather than runner? It seems like having the common approach at the above layer is needed anyway to make it efficient.

Reply:
For now I think this is to achieve consistency with synchronous `PartitionedDatasets`, not sure what you have in mind for runners but maybe we should discuss that separately? Unless you still see issues with the proposed approach